File: backend/app/api/charts.py
from typing import Any, List, Optional
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, Body
from sqlalchemy.orm import Session

from app.auth.deps import get_current_user, get_db
from app.models.models import Chart, User, Tag, chart_tags
from app.schemas.schemas import ChartCreate, ChartUpdate, Chart as ChartSchema, TagCreate, Tag as TagSchema

router = APIRouter()


@router.get("/", response_model=List[ChartSchema])
def read_charts(
    db: Session = Depends(get_db),
    skip: int = 0,
    limit: int = 100,
    tag: Optional[str] = None,
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Retrieve charts. Filter by tag if provided.
    """
    query = db.query(Chart).filter(Chart.user_id == current_user.id)
    
    # Filter by tag if provided
    if tag:
        tag_obj = db.query(Tag).filter(Tag.name == tag).first()
        if tag_obj:
            query = query.join(chart_tags).filter(chart_tags.c.tag_id == tag_obj.id)
    
    return query.order_by(Chart.created_at.desc()).offset(skip).limit(limit).all()


@router.post("/", response_model=ChartSchema)
def create_chart(
    *,
    db: Session = Depends(get_db),
    chart_in: ChartCreate,
    tags: Optional[List[str]] = Body(None),
    thread_id: Optional[str] = Body(None),
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Create new chart.
    """
    chart = Chart(
        user_id=current_user.id,
        title=chart_in.title,
        description=chart_in.description,
        query=chart_in.query,
        natural_language_query=chart_in.natural_language_query,
        provider=chart_in.provider,
        data=chart_in.data,
        vega_spec=chart_in.vega_spec,
        is_public=chart_in.is_public,
        last_refreshed_at=datetime.now(),
        execution_time_ms=0,  # This would be set based on actual execution time
        view_count=0,
    )
    db.add(chart)
    db.flush()  # Get the ID without committing transaction
    
    # Add tags if provided
    if tags:
        for tag_name in tags:
            # Get or create tag
            tag = db.query(Tag).filter(Tag.name == tag_name).first()
            if not tag:
                tag = Tag(name=tag_name)
                db.add(tag)
                db.flush()  # Get the ID without committing transaction
            
            # Associate tag with chart
            db.execute(chart_tags.insert().values(chart_id=chart.id, tag_id=tag.id))
    
    db.commit()
    db.refresh(chart)
    return chart


@router.get("/public", response_model=List[ChartSchema])
def read_public_charts(
    db: Session = Depends(get_db),
    skip: int = 0,
    limit: int = 100,
) -> Any:
    """
    Retrieve public charts.
    """
    return db.query(Chart).filter(Chart.is_public == True).offset(skip).limit(limit).all()


@router.get("/{chart_id}", response_model=ChartSchema)
def read_chart(
    *,
    db: Session = Depends(get_db),
    chart_id: int,
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Get chart by ID.
    """
    chart = db.query(Chart).filter(Chart.id == chart_id).first()
    if not chart:
        raise HTTPException(status_code=404, detail="Chart not found")
    if not chart.is_public and chart.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    return chart


@router.put("/{chart_id}", response_model=ChartSchema)
def update_chart(
    *,
    db: Session = Depends(get_db),
    chart_id: int,
    chart_in: ChartUpdate,
    tags: Optional[List[str]] = Body(None),
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Update a chart.
    """
    chart = db.query(Chart).filter(Chart.id == chart_id).first()
    if not chart:
        raise HTTPException(status_code=404, detail="Chart not found")
    if chart.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    
    update_data = chart_in.dict(exclude_unset=True)
    for field, value in update_data.items():
        setattr(chart, field, value)
    
    # Update tags if provided
    if tags is not None:
        # Clear existing tags
        db.execute(chart_tags.delete().where(chart_tags.c.chart_id == chart_id))
        
        # Add new tags
        for tag_name in tags:
            # Get or create tag
            tag = db.query(Tag).filter(Tag.name == tag_name).first()
            if not tag:
                tag = Tag(name=tag_name)
                db.add(tag)
                db.flush()  # Get the ID without committing transaction
            
            # Associate tag with chart
            db.execute(chart_tags.insert().values(chart_id=chart_id, tag_id=tag.id))
    
    db.add(chart)
    db.commit()
    db.refresh(chart)
    return chart


@router.delete("/{chart_id}", response_model=ChartSchema)
def delete_chart(
    *,
    db: Session = Depends(get_db),
    chart_id: int,
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Delete a chart.
    """
    chart = db.query(Chart).filter(Chart.id == chart_id).first()
    if not chart:
        raise HTTPException(status_code=404, detail="Chart not found")
    if chart.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    
    db.delete(chart)
    db.commit()
    return chart


@router.post("/{chart_id}/view")
def increment_chart_view(
    *,
    db: Session = Depends(get_db),
    chart_id: int,
    current_user: Optional[User] = Depends(get_current_user),
) -> Any:
    """
    Increment the view count for a chart.
    """
    chart = db.query(Chart).filter(Chart.id == chart_id).first()
    if not chart:
        raise HTTPException(status_code=404, detail="Chart not found")
    
    # Only allow viewing public charts or charts owned by the user
    if not chart.is_public and (not current_user or chart.user_id != current_user.id):
        raise HTTPException(status_code=403, detail="Not enough permissions")
    
    # Increment view count
    chart.view_count = chart.view_count + 1
    db.add(chart)
    db.commit()
    
    return {"status": "success", "view_count": chart.view_count}


@router.get("/{chart_id}/tags", response_model=List[TagSchema])
def read_chart_tags(
    *,
    db: Session = Depends(get_db),
    chart_id: int,
    current_user: User = Depends(get_current_user),
) -> Any:
    """
    Get all tags for a chart.
    """
    chart = db.query(Chart).filter(Chart.id == chart_id).first()
    if not chart:
        raise HTTPException(status_code=404, detail="Chart not found")
    
    if not chart.is_public and chart.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    
    return chart.tags

File: frontend/src/store/charts.ts
import { create } from 'zustand';
import { chartsAPI } from '@/utils/api';

export interface Chart {
  id: number;
  title: string;
  description?: string;
  query: string;
  natural_language_query: string;
  provider: string;
  data: Record<string, unknown>[];
  vega_spec: Record<string, unknown>;
  is_public: boolean;
  user_id: number;
  created_at: string;
  updated_at?: string;
  last_refreshed_at?: string;
  execution_time_ms?: number;
  view_count?: number;
  tags?: Tag[];
}

export interface Tag {
  id: number;
  name: string;
}

interface ChartsState {
  charts: Chart[];
  publicCharts: Chart[];
  currentChart: Chart | null;
  isLoading: boolean;
  error: string | null;
  fetchCharts: () => Promise<void>;
  fetchPublicCharts: () => Promise<void>;
  fetchChart: (id: number) => Promise<Chart>;
  createChart: (chartData: Partial<Chart>, tags?: string[]) => Promise<Chart>;
  updateChart: (id: number, chartData: Partial<Chart>) => Promise<Chart>;
  deleteChart: (id: number) => Promise<void>;
}

export const useChartsStore = create<ChartsState>((set, get) => ({
  charts: [],
  publicCharts: [],
  currentChart: null,
  isLoading: false,
  error: null,
  
  fetchCharts: async () => {
    set({ isLoading: true, error: null });
    
    try {
      const charts = await chartsAPI.getCharts();
      set({ charts, isLoading: false });
    } catch (error: unknown) {
      const err = error as { response?: { data?: { detail?: string } } };
      set({ 
        isLoading: false, 
        error: err.response?.data?.detail || 'Failed to fetch charts' 
      });
    }
  },
  
  fetchPublicCharts: async () => {
    set({ isLoading: true, error: null });
    
    try {
      const publicCharts = await chartsAPI.getPublicCharts();
      set({ publicCharts, isLoading: false });
    } catch (error: unknown) {
      const err = error as { response?: { data?: { detail?: string } } };
      set({ 
        isLoading: false, 
        error: err.response?.data?.detail || 'Failed to fetch public charts' 
      });
    }
  },
  
  fetchChart: async (id: number) => {
    set({ isLoading: true, error: null });
    
    try {
      const chart = await chartsAPI.getChart(id);
      set({ currentChart: chart, isLoading: false });
      return chart;
    } catch (error: any) {
      set({ 
        isLoading: false, 
        error: error.response?.data?.detail || 'Failed to fetch chart' 
      });
      throw error;
    }
  },
  
  createChart: async (chartData: Partial<Chart>, tags?: string[]) => {
    set({ isLoading: true, error: null });
    
    try {
      // Extract tags if they exist in chartData (for backward compatibility)
      const { tags: chartTags, ...chartDataWithoutTags } = chartData as Partial<Chart & { tags?: string[] }>;
      
      // Call API with chart data and tags separately, prioritizing passed tags parameter
      const tagsToUse = tags || chartTags;
      const newChart = await chartsAPI.createChart(chartDataWithoutTags, tagsToUse);
      
      set({ 
        charts: [...get().charts, newChart],
        isLoading: false 
      });
      return newChart;
    } catch (error: any) {
      set({ 
        isLoading: false, 
        error: error.response?.data?.detail || 'Failed to create chart' 
      });
      throw error;
    }
  },
  
  updateChart: async (id: number, chartData: Partial<Chart>) => {
    set({ isLoading: true, error: null });
    
    try {
      const updatedChart = await chartsAPI.updateChart(id, chartData);
      set({ 
        charts: get().charts.map(chart => 
          chart.id === id ? updatedChart : chart
        ),
        currentChart: updatedChart,
        isLoading: false 
      });
      return updatedChart;
    } catch (error: any) {
      set({ 
        isLoading: false, 
        error: error.response?.data?.detail || 'Failed to update chart' 
      });
      throw error;
    }
  },
  
  deleteChart: async (id: number) => {
    set({ isLoading: true, error: null });
    
    try {
      await chartsAPI.deleteChart(id);
      set({ 
        charts: get().charts.filter(chart => chart.id !== id),
        isLoading: false 
      });
    } catch (error: any) {
      set({ 
        isLoading: false, 
        error: error.response?.data?.detail || 'Failed to delete chart' 
      });
      throw error;
    }
  },
}))
